openedx/core/djangoapps/programs/utils.py

```diff
@@ -132,28 +133,120 @@ def get_display_category(program):
+    """Utility for gauging a user's progress towards program completion.
 
     Arguments:
         user (User): The user for which to find programs.
-        enrollments (list): The user's enrollments.
-
-    Returns:
-        list of serialized programs, ordered by most recent enrollment
+        enrollments (list): The user's active enrollments.
     """
-    programs = get_programs(user)
-
-    enrollments = sorted(enrollments, key=lambda e: e.created, reverse=True)
-    # enrollment.course_id is really a course key.
-    course_ids = [unicode(e.course_id) for e in enrollments]
+    def __init__(self, user, enrollments):
+        self.user = user
+        self.enrollments = sorted(enrollments, key=lambda e: e.created, reverse=True)
+
+        # enrollment.course_id is really a course key ಠ_ಠ
+        self.course_ids = [unicode(e.course_id) for e in self.enrollments]
+        self.engaged_programs = None
+        self.course_certs = None
+
+    @property
+    def programs(self):
+        """Derive a list of programs in which the given user is engaged.
+
+        Returns:
+            list of program dicts, ordered by most recent enrollment.
+        """
+        programs = get_programs(self.user)
+        flattened = flatten_programs(programs, self.course_ids)
+
+        self.engaged_programs = []
+        for course_id in self.course_ids:
+            for program in flattened.get(course_id, []):
+                if program not in self.engaged_programs:
+                    self.engaged_programs.append(program)
+
+        return self.engaged_programs
+
+    @property
+    def progress(self):
+        """Gauge a user's progress towards program completion.
+
+        Returns:
+            list of dict, each containing information about a user's progress
+                towards completing a program.
+        """
+        if self.engaged_programs is None:
+            self.programs  # pylint: disable=pointless-statement
+
+        self.course_certs = get_completed_courses(self.user)
+
+        progress = []
+        for program in self.engaged_programs:
+            completed, in_progress, not_started = [0] * 3
 
-    flattened = flatten_programs(programs, course_ids)
-
-    engaged_programs = []
-    for course_id in course_ids:
-        for program in flattened.get(course_id, []):
-            if program not in engaged_programs:
-                engaged_programs.append(program)
-
-    return engaged_programs
+            for course_code in program['course_codes']:
+                if self._is_complete(course_code):
+                    completed += 1
+                elif self._is_in_progress(course_code):
+                    in_progress += 1
+                else:
+                    not_started += 1
+
+            progress.append({
+                'id': program['id'],
+                'completed': completed,
+                'in_progress': in_progress,
+                'not_started': not_started,
+            })
```

Comment:
@rlucioni suggestion.  In the next milestone it will be necessary to display the completed / in progress / not started information correlated with the individual courses/runs.  Instead of returning counts by status in this API,  would it make sense to return tuples of course codes?  One can do `len()` on those values to get the same counts being returned here, but it will also be possible to do the richer displays without API changes.